cli/deno_dir.rs

```diff
@@ -397,26 +405,233 @@ fn get_source_code_async(
   debug!("is remote but didn't find module");
 
   // not cached/local, try remote
-  Either::B(fetch_remote_source_async(&module_name, &filename).and_then(
-    move |maybe_remote_source| match maybe_remote_source {
-      Some(output) => Ok(output),
-      None => Err(DenoError::from(std::io::Error::new(
-        std::io::ErrorKind::NotFound,
-        format!("cannot find remote file '{}'", &filename),
-      ))),
-    },
-  ))
+  Either::B(
+    fetch_remote_source_async(deno_dir, &module_name, &filename).and_then(
+      move |maybe_remote_source| match maybe_remote_source {
+        Some(output) => Ok(output),
+        None => Err(DenoError::from(std::io::Error::new(
+          std::io::ErrorKind::NotFound,
+          format!("cannot find remote file '{}'", &filename),
+        ))),
+      },
+    ),
+  )
 }
 
 #[cfg(test)]
 /// Synchronous version of get_source_code_async
 /// This function is deprecated.
 fn get_source_code(
+  deno_dir: &DenoDir,
   module_name: &str,
   filename: &str,
   use_cache: bool,
 ) -> DenoResult<ModuleMetaData> {
-  tokio_util::block_on(get_source_code_async(module_name, filename, use_cache))
+  tokio_util::block_on(get_source_code_async(
+    deno_dir,
+    module_name,
+    filename,
+    use_cache,
+  ))
+}
+
+/// Fetch local or cached source code.
+/// This is a recursive operation if source file has redirection.
+/// It will keep reading filename.meta.json for information about redirection.
+/// module_initial_source_name would be None on first call,
+/// and becomes the name of the very first module that initiates the call
+/// in subsequent recursions.
+/// AKA if redirection occurs, module_initial_source_name is the source path
+/// that user provides, and the final module_name is the resolved path
+/// after following all redirections.
+fn fetch_local_source(
+  deno_dir: &DenoDir,
+  module_name: &str,
+  filename: &str,
+  module_initial_source_name: Option<String>,
+) -> DenoResult<Option<ModuleMetaData>> {
+  let p = Path::new(&filename);
+  let source_metadata = get_source_code_metadata(&filename);
+  // If source metadata says that it would redirect elsewhere,
+  // (meaning that the source file might not exist; only .meta.json is present)
+  // Abort reading attempts to the cached source file and and follow the redirect.
+  if let Some(redirect_to) = source_metadata.redirect_to {
+    // e.g.
+    // module_name https://import-meta.now.sh/redirect.js
+    // filename /Users/kun/Library/Caches/deno/deps/https/import-meta.now.sh/redirect.js
+    // redirect_to https://import-meta.now.sh/sub/final1.js
+    // real_filename /Users/kun/Library/Caches/deno/deps/https/import-meta.now.sh/sub/final1.js
+    // real_module_name = https://import-meta.now.sh/sub/final1.js
```

Comment:
I want to really try to keep DenoDir out of the arguments here. My observation is that the argument `filename` already contains all the info we need about the location of the `DenoDir`. If only we had a function like this:
```
fn find_gen_root_from_cache_path(filename: &str) -> PathBuf
```
It seems a bit tricky - because I think in the case of `file:///` modules, the `filename` would not actually point into the cache but to location on disk. So maybe this is something we have to do later.

That said, please consider if this would be possible to remove the DenoDir argument here. It might be too much work...

Reply:
Working on tests to examine goal functionalities. Would look into this once done.

Reply:
Actually I'm kind of worried that such deduction might be brittle: just using filename is not enough since it is totally possible that there are user create path segments like `deps/https` or whatever. Combining with `module_name` might work but I really feel it might be needlessly complicated... especially considering we might want to refactor deno_dir soon...

That being said, I'm more inclined to make it a follow-up PR if we really want to do this.

Reply:
Ignore my require then. Thanks.